### ISSUE TITLE
qtdeploy generated package issues

### ISSUE THREAD

**Issue by akamensky** [therecipe/qt#184]
I've comparing the application I just finished porting to go with the version built with qt creator. Original project used C++ only for GUI part, the rest was written in Go and exported with CGo to be linked with GUI. 

There are a few instantly noticeable issues after migration:

1. Binary size is more than double of original app (4440836 bytes of original vs 10807416 of ported to use these bindings).

2. For some reason it adds print support framework from Qt, even though the binary is not linked against it (the only used frameworks are - Core, Gui and Widgets)

Also some small annoyances - not clear how to provide Info.plist and other files. E.g. this project uses a helper tool (executed with exec.Command) to do operations requiring privilege escalation, in Qt Creator it is possible to specify extra files that will be included in deploy step, here it does not seem to be possible at the moment. To name a few - Info.plist, `.icns` bundle, App icon (separate from .icns), external tools.

**Comment by therecipe** [comment]
>Binary size is more than double of original app (4440836 bytes of original vs 10807416 of ported to use these bindings).

Yes, this is probably dead code (`qtminimal` is not really aggressive and only sorts out those functions  that are obviously not needed)

Did you used `qtdeploy` or `go build`?
Maybe you can make it a bit smaller by stripping the binary with the `-s` and `-w` flags.
And on mac you can probably also use `strip` directly.

>For some reason it adds print support framework from Qt, even though the binary is not linked against it (the only used frameworks are - Core, Gui and Widgets)

This could be because of `qtmoc`, can you check if the generated `cgo*` moc files link to the printsupport framework?

>Also some small annoyances - not clear how to provide Info.plist and other files. E.g.
 
Yes, I haven't really looked into that yet. For mobile targets though it can be somewhat done with the automatically created subfolders called `android`, `ios`, `sailfish`, ...

**Comment by akamensky** [comment]
> Did you used qtdeploy or go build?

I used qtdeploy. Stripping does not decrease the size at all (even unsafe stripping), so I guess it is either dead code gets in somehow, or GCC/Clang optimization flags (there are plenty of `.cpp` to build in qt part judging from the structure). Although original project uses default optimizations (`-O2`), which I presume also left by default for any CGo compilations.

> can you check if the generated cgo* moc files link to the printsupport framework?

Checked, there is no mention of printsupport anywhere in generated files. Also I guess if it was, then the relevant framework would be linked to the final binary, but it is not (checked with `otool -L` on Mac). The only really linked dependencies are Gui, Core and Widgets.

**Comment by therecipe** [comment]
>Although original project uses default optimizations (-O2), which I presume also left by default for any CGo compilations.

Yes, `-O2` is used but I think than it's just dead code which is not sorted out by the go compiler.
Go 1.6 to 1.7 for example reduced the binary size around 30% for me on average, so there is probably still much to be gained from their improvements on the compiler.

>For some reason it adds print support framework from Qt, even though the binary is not linked 
>The only really linked dependencies are Gui, Core and Widgets.

Oh, missed this sorry.
Then it's probably `macdeployqt` that copies more than is necessary.
`macdeployqt` is also the reason, why I couldn't update to 5.7.1 yet :(
https://bugreports.qt.io/browse/QTBUG-57265

**Comment by akamensky** [comment]
@therecipe Sorry for the slow follow up.

> Then it's probably macdeployqt that copies more than is necessary

That seems to be correct, building app using C++ in Qt Creator produces same result (print support framework being included). Though it can be remove since the app is not linked against this library.

> Go 1.6 to 1.7 for example reduced the binary size around 30%

It actually appears to be dead code for getting in there for some reason. Not sure how to look for the dead code blocks (or if it is even possible to find them somehow). Would be nice if someone could get a look at this rather annoying issue.

**Comment by therecipe** [comment]
>@therecipe Sorry for the slow follow up.

no problem :)

>It actually appears to be dead code for getting in there for some reason. Not sure how to look for the dead code blocks (or if it is even possible to find them somehow). Would be nice if someone could get a look at this rather annoying issue.

there are some changes coming after the update to 5.8, which will reduce the generated amount of code in general. But I'm not sure whether that's going to reduce your binary size by 50%, it's probably in the range of 10-15% at best.

**Comment by akamensky** [comment]
@therecipe Just reviewing my issues here (checking if I could already use this project for my app). Seems to the size of app is still increased when using moc classes.

I've used example internal/examples/widgets/systray code for comparison:

Unedited code compiled as is produces binary:
```
-rwxr-xr-x  1 akamensky  staff  2393896 Jun 12 22:15 main
```

Edited code to disable mocking (see code example below):
```
-rwxr-xr-x  1 akamensky  staff  1889544 Jun 12 22:25 main
```

It still looks like too much overhead which does not happen with custom slots in native C++ code.

<details markdown="1"> <summary markdown="0">Edited code:</summary>

    package main
    
    import (
    	"os"
    
    	"github.com/therecipe/qt/widgets"
    )
    
    //type QSystemTrayIconWithCustomSlot struct {
    //	widgets.QSystemTrayIcon
    //
    //	_ func() `slot:"triggerSlot"`
    //}
    
    func main() {
    	widgets.NewQApplication(len(os.Args), os.Args)
    
    	var (
    		widget       = widgets.NewQWidget(nil, 0)
    		widgetLayout = widgets.NewQVBoxLayout()
    	)
    	widget.SetLayout(widgetLayout)
    
    //	var systray = NewQSystemTrayIconWithCustomSlot(nil)
    	var systray = widgets.NewQSystemTrayIcon(nil)
    	systray.SetIcon(widget.Style().StandardIcon(widgets.QStyle__SP_MessageBoxCritical, nil, nil))
    
    	var systrayMenu = widgets.NewQMenu(nil)
    	systrayMenu.AddAction("first action")
    	systrayMenu.AddAction("second action")
    	systray.SetContextMenu(systrayMenu)
    
    	systray.Show()
    
    	//works
    	var buttonMain = widgets.NewQPushButton2("call from main thread", nil)
    	buttonMain.ConnectClicked(func(bool) {
    		systray.ShowMessage("title", "main thread message", widgets.QSystemTrayIcon__Information, 5000)
    	})
    	widgetLayout.AddWidget(buttonMain, 0, 0)
    
    	//fails
    	var buttonOther = widgets.NewQPushButton2("call from other thread", nil)
    	buttonOther.ConnectClicked(func(bool) {
    		go func() {
    			systray.ShowMessage("title", "other thread message", widgets.QSystemTrayIcon__Information, 5000)
    		}()
    	})
    	widgetLayout.AddWidget(buttonOther, 0, 0)
    
    	//works
    //	var buttonSlot = widgets.NewQPushButton2("call from other thread with slot", nil)
    //	systray.ConnectTriggerSlot(func() {
    //		systray.ShowMessage("title", "other thread message with slot", widgets.QSystemTrayIcon__Information, 5000)
    //	})
    //	buttonSlot.ConnectClicked(func(bool) {
    //		go func() {
    //			systray.TriggerSlot()
    //		}()
    //	})
    //	widgetLayout.AddWidget(buttonSlot, 0, 0)
    
    	widget.Show()
    
    	widgets.QApplication_Exec()
    }

</details>

**Comment by therecipe** [comment]
Yeah, that's to much for a simple class + slot.
I was planning to work on some moc related changes anyway, so I will look into this as well when I'm at it.

**Comment by akamensky** [comment]
Also seems for every slotted class it produces a bunch of dead code. So in my app with 10+ slotted classes it ends up doubling final executable in size.

**Comment by therecipe** [comment]
Sorry, it took a little longer.
I looked into it and figured that there is also some other major issue, beside the size increase of the binary.

The problem is that if you create a moc struct, you will start to link against a whole bunch of probably unnecessary modules.

```go
package main

import (
	_ "github.com/therecipe/qt/core"
)

func main() {

}
```

This will compile to 664K and links against 

```
@rpath/QtGui.framework/Versions/5/QtGui
@rpath/QtCore.framework/Versions/5/QtCore
@rpath/QtSvg.framework/Versions/5/QtSvg
@rpath/QtWidgets.framework/Versions/5/QtWidgets
```

the final size of the *.app is 20M

```go
package main

import (
	"github.com/therecipe/qt/core"
)

type object struct {
	core.QObject
}

func main() {

}
```

This however complies to 1,4M and links against

```
@rpath/QtMultimediaWidgets.framework/Versions/5/QtMultimediaWidgets
@rpath/QtMultimedia.framework/Versions/5/QtMultimedia
@rpath/QtNetwork.framework/Versions/5/QtNetwork
@rpath/QtCore.framework/Versions/5/QtCore
@rpath/QtGui.framework/Versions/5/QtGui
@rpath/QtWidgets.framework/Versions/5/QtWidgets
@rpath/QtDesigner.framework/Versions/5/QtDesigner
@rpath/QtXml.framework/Versions/5/QtXml
@rpath/QtQuickWidgets.framework/Versions/5/QtQuickWidgets
@rpath/QtQuick.framework/Versions/5/QtQuick
@rpath/QtQml.framework/Versions/5/QtQml
@rpath/QtDBus.framework/Versions/5/QtDBus
@rpath/QtTest.framework/Versions/5/QtTest
@rpath/QtSvg.framework/Versions/5/QtSvg
```

and the final size of the *.app is  37M


However the good news are, that both issue are related and can be theoretically solved by making some changes to qtmoc and qtminimal.
I will fix this asap.
